File: DataStructure/LinkedList/max_subarray.py
from linked_list import Node
from linked_list import LinkedList

"""
Problem Statement
You have been given an array containg numbers. Find and return the largest sum in a contiguous subarray within the input array.

Example 1:

arr= [1, 2, 3, -4, 6]
The largest sum is 8, which is the sum of all elements of the array.
Example 2:

arr = [1, 2, -5, -4, 1, 6]
The largest sum is 7, which is the sum of the last two elements of the array.
"""
def max_sum_subarray(arr):
    """
    :param - arr - input array
    return - number - largest sum in contiguous subarry within arr
    Author: Zhijin Li (lizhijin1993@gmail.com)
    """
    current_sum = arr[0]
    max_sum = arr[0]
    for num in arr[1:]:
    	current_sum = max(num, current_sum + num)
    	max_sum = max(current_sum, max_sum)
    return max_sum

def test_function(test_case):
    arr = test_case[0]
    solution = test_case[1]
    
    output = max_sum_subarray(arr)
    if output == solution:
        print("Pass")
    else:
        print("Fail")

# Test Case 1
arr= [1, 2, 3, -4, 6]
solution= 8 # sum of array

test_case = [arr, solution]
test_function(test_case)

# Test Case 2
arr = [1, 2, -5, -4, 1, 6]
solution = 7   # sum of last two elements

test_case = [arr, solution]
test_function(test_case)

# Test Case 3
arr = [-12, 15, -13, 14, -1, 2, 1, -5, 4]
solution = 18  # sum of subarray = [15, -13, 14, -1, 2, 1]

test_case = [arr, solution]
test_function(test_case)

File: DataStructure/LinkedList_and_Array/maxSubarray.js
/*
Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

Example:

Input: [-2,1,-3,4,-1,2,1,-5,4],
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.
*/

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {

    if(nums.length === 0){
        return 0;
    }
    let currSum = 0;
    let maxSum = nums[0];

    for(let i=0; i<nums.length;i++){
        if(nums[i]+ currSum < 0){
            currSum = 0;
            maxSum = Math.max(maxSum, nums[i]);
        }else{
            currSum += nums[i];
            maxSum = Math.max(maxSum, currSum);
        }
    }
    return maxSum;
}

/*
Complexity Analysis

Time complexity : O(N) since it's one pass along the array.

Space complexity : O(1), since it's a constant space solution.
*/

console.log(maxSubArray([-2,1,-3,4,-1,2,1,-5,4]));